packages/cdktf-cli/lib/server/WatchClient.ts

```diff
@@ -0,0 +1,395 @@
+    if (!isDeepEqual(previousState, this.state)) {
+      this.subscribers.forEach((sub) => sub(this.state as WatchState));
+    }
+  }
+
+  private async runSynth() {
+    this.updateState({ status: "SYNTHESIZING", error: undefined });
+    try {
+      const stacks = (
+        await SynthStack.synth(this.synthCommand, this.targetDir, true, "watch")
+      ).map<Stack>((stack) => {
+        return {
+          ...stack,
+          json: JSON.parse(stack.content) as TerraformJson,
+        };
+      });
+      this.updateState({
+        stacks,
+      });
+      if (this.needsInit) await this.queueAction("INIT", true);
+    } catch (e) {
+      throw new RecoverableError(e.errorOutput || e.message, "SYNTHESIZING");
+    }
+  }
+
+  private async runInit() {
+    this.updateState({ status: "INITIALIZING" });
+    const terraform = await this.getTerraform();
+    await terraform.init();
+    this.needsInit = false;
+  }
+
+  private async runDeploy() {
+    const { workingDirectory, name } = await this.getTargetStack();
+    console.log({ workingDirectory });
+    const newTargetStackHash = hashPath(workingDirectory);
+    if (this.lastTargetStackHash === newTargetStackHash) {
+      console.log(
+        `skipping deploy because hash ("${this.lastTargetStackHash}") of target stack "${name}" did not change for working directory ${workingDirectory}`
+      );
+      return;
+    } else {
+      console.log(
+        `continuing deploy because new hash ("${newTargetStackHash}") of working directory "${workingDirectory}" differs from previous hash "${this.lastTargetStackHash}"`
+      );
+    }
+
+    this.updateState({ status: "DEPLOYING" });
+    const terraform = await this.getTerraform();
+
+    // remove previous errors
+    // we do this here as deploy will update the state already while
+    // in progress and we don't want to see the error while successful
+    // things already start to happen
+    this.updateState({
+      error: undefined,
+    });
+    try {
+      if (terraform instanceof TerraformCloud) {
+        // plan for Terraform Cloud remote execution also uploads the Terraform code
+        const { planFile } = await terraform.plan(false); // false = deploy
+        await terraform.deploy(planFile, this.handleTerraformOutput.bind(this));
+      } else if (terraform instanceof TerraformCli) {
+        // skip the plan and refreshing to save time ⏱
+        const NO_PLAN_FILE = "";
+        await terraform.deploy(
+          NO_PLAN_FILE,
+          this.handleTerraformOutput.bind(this),
+          ["-refresh=false"]
+        );
+      }
+      // deployment was successful -> update hash
+      this.lastTargetStackHash = newTargetStackHash;
+    } catch (e) {
+      // When new providers where added or the stack name might've
+      // been changed (when using just a single stack) we need to
+      // re-init Terraform. We'll do this proactively on any error.
+      this.needsInit = true;
```

Comment:
Because currently after a recoverable error the WatchClient will be IDLE until a synth or a deploy is triggered. It will then run an init after the synth but before another deploy. So in the case of synth the status would be SYNTHESIZING and then DEPLOYING while needsInit needs to be retained until the synth finished.
But yeah, I'd love to refactor this into a proper state machine at some point. I also think this is ugly.